### PR TITLE
Optimize calculate_data and redact_data()

### PR DESCRIPTION
**Pull request by Lash-L**
redact data likely needs more work. Gonna let my instance run for a while and then try it. With limited data it does show some pretty large improvements.

There were some merge conflicts for calculate_data and update_advertisement. So if you could just give those a extra lookover that would probably be good!

### CODE REVIEW

File: custom_components/bermuda/coordinator.py [hunk 1]
```diff
@@ -1282,8 +1282,18 @@ def redact_data(self, data):
             self.redaction_list_update()
         if isinstance(data, str):
             # the end of the recursive wormhole, do the actual work:
-            for find, fix in self.redactions.items():
-                data = re.sub(find, fix, data, flags=re.IGNORECASE)
+            if ":" in data:
+                if data not in self.redactions:
+                    if data.upper() not in self.redactions:
```
Comment (agittins) [inline custom_components/bermuda/coordinator.py:1287]: Ha! Brilliant!

**Review by agittins — COMMENTED** [review]
This is great stuff :D

I'm also thinking that we should:

- in `redaction_list_update()`:
  - store a (errr... static local?) stamp of MONOTONIC_TIME(), and if the last one was more than... maybe several hours? Then re-initialise list so we don't end up iterating over fixes for addresses (and associated matches) that have long since been purged from the system (ie, when a user does two diags but quite some time apart). There's probably a trade-off between cache miss/size cost and regeneration-time.
  - (maybe/or) schedule a job in hass to purge the redactions[] after a certain time, to free up the memory I assume they'd take up. Maybe it's not big enough to matter.
  - never apply `.upper()` to `address` (since we always normalise the address to lowercase) and always `lower()` any non-known-lowered address when adding/searching keys in redactions.

- Then in `redact_data()` we can remove the case-checking variants, and always look for `redactions[thing.lower()]` / `if thing.lower() in redactions` or `if find in data.lower()`.

Also, maybe we can ditch the re.sub altogether and use python's string.replace() since it might be a lot faster and we're not using the power of regex at all, really. Notes should be added to the update_redactions info to explain that the keys are not regex, but lower() forms of literal strings to match.

Also...

[This comment](https://stackoverflow.com/a/35241298) is an interesting idea that might apply to our use-case:
> If the replacements do not change the string size, you can speed up things a lot by using bytearray / buffers or even mmap and modify the data in-place. (re.sub() and string.replace and endstring += line cause that a lot of memory is copyied around.)

File: custom_components/bermuda/bermuda_device_scanner.py [hunk 2]
```diff
@@ -105,19 +121,20 @@ def update_advertisement(self, scandata: BluetoothScannerDevice):
...
+        self.name: str = scanner.name
+        self.area_id: str = self.scanner_device.area_id
         self.area_name = self.scanner_device.area_name
-        new_stamp = None
-
+        new_stamp: float | None
         # Only remote scanners log timestamps here (local usb adaptors do not),
-        if hasattr(scandata.scanner, "_discovered_device_timestamps"):
+        if scanner.source in self.cached_remote_scanners or isinstance(scanner, BaseHaRemoteScanner):
+            self.cached_remote_scanners.add(scanner.source)
```
[the remark this answers is not in the document]
Comment (Lash-L) [reply custom_components/bermuda/bermuda_device_scanner.py:125]: Both! hasattr is known for being pretty slow. BaseHaRemoteScanner is what has the _discovered_device_timestamps property. I guess it is possible if someone is doing some weird overloading somewhere? But definitely no scanners in core are like that.

File: custom_components/bermuda/bermuda_device_scanner.py [hunk 3]
```diff
@@ -387,9 +392,13 @@ def calculate_data(self):
...
                 # Looks valid enough, add the current reading to the interval log
                 self.hist_distance_by_interval.insert(0, self.rssi_distance_raw)
+            dist_count = len(self.hist_distance_by_interval)
 
             # trim the log to length
-            del self.hist_distance_by_interval[self.options.get(CONF_SMOOTHING_SAMPLES) :]
+            if self.smoothing_samples < dist_count:
+                del self.hist_distance_by_interval[self.smoothing_samples :]
+                # It should only ever need to remove one
+                dist_count -= 1
```
[the remark this answers is not in the document]
Comment (Lash-L) [reply custom_components/bermuda/bermuda_device_scanner.py:401]: As far as I am aware, del does not do a len check. And arithmetic operation should be faster than deleting nothing.

That being said, I assume most devices will hit the smoothing samples number and therefore always have to delete something, so maybe it makes sense to remove the if check.

File: custom_components/bermuda/bermuda_device_scanner.py [hunk 4]
```diff
@@ -454,3 +455,7 @@ def to_dict(self):
                 continue
             out[var] = val
         return out
+
+    def __repr__(self) -> str:
+        """Help debugging by giving it a clear name instead of empty dict."""
+        return self.address
```
[the remark this answers is not in the document]
Comment (Lash-L) [reply custom_components/bermuda/bermuda_device_scanner.py:461]: No, repr is just the representation of the object. If I modified str, it likely would, but we wouldn't be calling repr in that bit of code.

**Comment by Lash-L** [comment]
> I'm also thinking that we should:
> 
> * in `redaction_list_update()`:
>   
>   * store a (errr... static local?) stamp of MONOTONIC_TIME(), and if the last one was more than... maybe several hours? Then re-initialise list so we don't end up iterating over fixes for addresses (and associated matches) that have long since been purged from the system (ie, when a user does two diags but quite some time apart). There's probably a trade-off between cache miss/size cost and regeneration-time.
>   * (maybe/or) schedule a job in hass to purge the redactions[] after a certain time, to free up the memory I assume they'd take up. Maybe it's not big enough to matter.
>   * never apply `.upper()` to `address` (since we always normalise the address to lowercase) and always `lower()` any non-known-lowered address when adding/searching keys in redactions.
> * Then in `redact_data()` we can remove the case-checking variants, and always look for `redactions[thing.lower()]` / `if thing.lower() in redactions` or `if find in data.lower()`.
> 
> Also, maybe we can ditch the re.sub altogether and use python's string.replace() since it might be a lot faster and we're not using the power of regex at all, really. Notes should be added to the update_redactions info to explain that the keys are not regex, but lower() forms of literal strings to match.

* in `redaction_list_update()`:
   
  * [ ] store a (errr... static local?) stamp of MONOTONIC_TIME(), and if the last one was more than... maybe several hours? Then re-initialise list so we don't end up iterating over fixes for addresses (and associated matches) that have long since been purged from the system (ie, when a user does two diags but quite some time apart). There's probably a trade-off between cache miss/size cost and regeneration-time.
  * [x] (maybe/or) schedule a job in hass to purge the redactions[] after a certain time, to free up the memory I assume they'd take up. Maybe it's not big enough to matter.
  * [x] never apply `.upper()` to `address` (since we always normalise the address to lowercase) and always `lower()` any non-known-lowered address when adding/searching keys in redactions.
* [x] Then in `redact_data()` we can remove the case-checking variants, and always look for `redactions[thing.lower()]` / `if thing.lower() in redactions` or `if find in data.lower()`.
* [x] ditch the re.sub

**Comment by Lash-L** [comment]
Ready for another lookover @agittins 

Redactions are deleted every 8 hours.

File: custom_components/bermuda/bermuda_device_scanner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (agittins) [reply custom_components/bermuda/bermuda_device_scanner.py:125]: Great, cheers!

**Comment by agittins** [comment]
Just a heads-up that I've had to make some changes to get some bugfixes out, so I've rebased your PR against main, I hope I've done that right :-/ There might be some areas (particularly in coordinator) where I've made changes to main that might need careful merging - I'm not 100% sure, I'm still getting to grips with handling merge conflicts (only been mucking around with various VCSs since the early 2000's, can't go too hard on myself! 😅 )

Re the `purging`, I think there are job/task types that should cancel automatically when the config entry is blatted - I think the ones you attach via the config entry itself might fit that - but I didn't find any that let you schedule future jobs when I was altering the device registry refresh thingie, so just used a timestamp.

**Comment by Lash-L** [comment]
Your merge looks fine! 

Yeah there is probably a better way to handle it, but honestly this seems to be fine imo. I think the tests were just being annoying

**Comment by agittins** [comment]
Whew! Sorry that took a long time for me to move on.

I got a bit wedged with stuff that I wanted finessed and got messed up with trying to merge with the latest changes, so I decided to just push the changes I wanted and get it merged :-)

Hopefully the PR attribution etc all works correctly and I haven't backed out anything in your PR while getting confused resolving merge conflicts. I have made some opinionated choices re using len() and possibly insert, but I think I've kept the core of what you have been working on, and the redaction stuff especially is amazing!

**Pull request merged**